File: 733-flood-fill/733-flood-fill.java
import java.awt.Point;

class Solution {
    
    // let m be length, n be height
    static int m;
    static int n;
    static int startcol;
    
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        Queue<Point> queue = new LinkedList<Point>();
        Point start = new Point(sr, sc);
        m = image[0].length;
        n = image.length;
        
        startcol = image[sr][sc];
        
        if (startcol == color) {
            return image;
        }
        
        queue.add(start);
        while (!queue.isEmpty()) {
            Point curr = queue.poll();
            int curr_len = (int) curr.getX();
            int curr_height = (int) curr.getY();
            
            image[curr_len][curr_height] = color;
            
            if (isValid(curr_len, curr_height - 1)) {
                if (image[curr_len][curr_height - 1] == startcol) {
                    Point new_pt = new Point(curr_len, curr_height - 1);
                    queue.add(new_pt);
                } 
            }
            
            if (isValid(curr_len, curr_height + 1)) {
                if (image[curr_len][curr_height + 1] == startcol) {
                    Point new_pt = new Point(curr_len, curr_height + 1);
                    queue.add(new_pt);
                } 
            } 
            
            if (isValid(curr_len - 1, curr_height)) {
                if (image[curr_len - 1][curr_height] == startcol) {
                    Point new_pt = new Point(curr_len - 1, curr_height);
                    queue.add(new_pt);
                } 
            } 
            
            if (isValid(curr_len + 1, curr_height)) {
                if (image[curr_len + 1][curr_height] == startcol) {
                    Point new_pt = new Point(curr_len + 1, curr_height);
                    queue.add(new_pt);
                } 
            }
        }
        return image;
        
    }
    
    public boolean isValid(int sr, int sc) {
        if (sr >= n) {
            return false;
        } else if (sc >= m) {
            return false;
        } else if (sr < 0 || sc < 0) {
            return false;
        } else {
            return true;
        }
    }
}

File: 733-flood-fill/733-flood-fill.py
class Solution(object):
    
    grid = []
    rows = -1;
    cols = -1;
    target = -1;
    original = -1;
    
    def floodFill(self, image, sr, sc, color):
        self.grid = image
        self.rows = len(image)
        self.cols = len(image[0])
        
        if (self.rows == 0 or self.cols == 0):
            return image
        
        self.target = color
        self.original = image[sr][sc]
        self.dfs(sr,sc);
        return self.grid
        
    def dfs(self, x, y):
        if (x < 0 or x >= self.rows or y < 0 or y >= self.cols):
            return
        elif (self.grid[x][y] == self.target or self.grid[x][y] != self.original):
            return
        else: 
            self.grid[x][y] = self.target
            self.dfs(x + 1, y)
            self.dfs(x - 1, y)
            self.dfs(x, y + 1)
            self.dfs(x, y - 1)